lib/vrl/compiler/src/expression/abort.rs

```diff
@@ -1,24 +1,72 @@
 use std::fmt;
 
+use diagnostic::{DiagnosticError, Label, Note, Urls};
+use parser::ast::Node;
+
 use crate::{
     expression::{ExpressionError, Resolved},
+    value::Kind,
     Context, Expression, Span, State, TypeDef,
 };
 
-#[derive(Debug, Clone, Copy, PartialEq)]
+use super::Expr;
+
+#[derive(Debug, Clone, PartialEq)]
 pub struct Abort {
     span: Span,
+    message: Option<Box<Expr>>,
 }
 
 impl Abort {
-    pub fn new(span: Span) -> Abort {
-        Abort { span }
+    pub fn new(span: Span, message: Option<Node<Expr>>, state: &State) -> Result<Self, Error> {
+        if let Some(node) = message {
+            let (expr_span, expr) = node.take();
+            let type_def = expr.type_def(state);
+
+            if type_def.is_fallible() {
+                Err(Error {
+                    variant: ErrorVariant::FallibleExpr,
+                    expr_span,
+                })
+            } else if !type_def.is_bytes() {
+                Err(Error {
+                    variant: ErrorVariant::NonString(type_def.kind()),
+                    expr_span,
+                })
```

Comment:
We have a `value::Error::Expected` variant that covers this, and unifies the returned error.